File: algorithms/counting-sort/c/counting.c
// Counting sort in C programming
#include <time.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void countingSort(long int arr[], int n)
{
  long int max = arr[0];
  for (int i = 1; i < n; i++)
  {
    if (arr[i] > max)
    {
      max = arr[i];
    }
  }

  long int *count = (long int *)calloc(max + 1, sizeof(long int));
  long int *output = (long int *)malloc(n * sizeof(long int));

  if (!count || !output)
  {
    printf("Erro de alocação de memória!\n");
    return;
  }

  for (int i = 0; i < n; i++)
  {
    count[arr[i]]++;
  }

  for (int i = 1; i <= max; i++)
  {
    count[i] += count[i - 1];
  }

  for (int i = n - 1; i >= 0; i--)
  {
    output[count[arr[i]] - 1] = arr[i];
    count[arr[i]]--;
  }

  for (int i = 0; i < n; i++)
  {
    arr[i] = output[i];
  }

  free(count);
  free(output);
}

void printArr(long int a[], int n)
{
  for (int i = 0; i < n; ++i)
    printf("%ld ", a[i]);

  printf("\n");
}
int main()
{
  FILE *file;
  long int number; // Use um tipo adequado para seus números
  long int *numbers = NULL;
  long int capacity = 1000; // Tamanho inicial do array
  long int size = 0;

  file = fopen("./number-generator/numbers.txt", "rb");
  if (NULL == file)
  {
    printf("Unable to open file\n");
    exit(-1);
  }

  numbers = (long int *)malloc(capacity * sizeof(long int)); // Alocar espaço inicial
  if (numbers == NULL)
  {
    printf("Unable to allocate memory\n");
    return 1;
  }

  while (fscanf(file, "%ld,", &number) != EOF)
  {
    if (size >= capacity)
    { // Realocar em incrementos maiores
      capacity *= 2;
      numbers = (long int *)realloc(numbers, capacity * sizeof(long int));
      if (numbers == NULL)
      {
        printf("Unable to reallocate memory\n");
        return 1;
      }
    }
    numbers[size] = number;
    size++;
  }

  fclose(file);

  printf("\nIniciando Execucao do Counting Sort em C: \n");
  countingSort(numbers, size);
  /*   printArr(numbers, size); */

  free(numbers);
}

File: algorithms/counting-sort/go/counting.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func max(values []int) int {
	//assign the first element equal to min
	max := values[0] //assign the first element equal to max
	for _, number := range values {
		if number > max {
			max = number
		}
	}
	return max
}

func coutingSort(array []int) []int {
	size := len(array)
	bigger := max(array) + 1
	outputPosition := 0

	var output []int = make([]int, size)
	var count []int = make([]int, bigger)

	for i := 0; i < size; i++ {
		count[array[i]] += 1
	}

	for i := 0; i < len(count); i++ {
		if count[i] != 0 {
			for count[i] > 0 {
				output[outputPosition] = i
				outputPosition += 1
				count[i] -= 1
			}
		}

	}

	return output
}

func main() {

	numbersFile, err := os.ReadFile("./number-generator/numbers.txt")
	if err != nil {
		log.Fatal(err)
	}

	var numbers []int
	numbersString := strings.Split(string(numbersFile), ",")

	for _, number := range numbersString {
		n, err := strconv.Atoi(number)
		if err != nil {
			log.Fatal(err)
		}

		numbers = append(numbers, n)
	}

	for _, arg := range os.Args[1:] {
		if n, err := strconv.Atoi(arg); err == nil {
			numbers = append(numbers, n)
		}
	}

	fmt.Printf("\nIniciando Execucao do Counting Sort em Go: \n\n")
	numbers = coutingSort(numbers)
	/* fmt.Printf("%v\n", numbers) */

}
